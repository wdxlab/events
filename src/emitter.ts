import { BaseEmitter, EmitterOptions, Filter } from './baseEmitter';
import { AggregateError } from './aggregateError';

export class Emitter<TArg> extends BaseEmitter<TArg, unknown[]> {
  constructor(options?: EmitterOptions<TArg, unknown[]>) {
    super(options);

    this.sourceEmitter?.on((arg) => {
      const result = this.emitInt(arg);
      if (result.length) {
        throw new AggregateError(result);
      }
    });
  }

  emit(arg: TArg): unknown[] {
    if (this.isReadOnly) {
      throw new Error('Emitter is readonly');
    }

    return this.emitInt(arg);
  }

  protected emitInt(arg: TArg): unknown[] {
    if (this.filter && !this.filter(arg)) {
      return [];
    }

    const errors: unknown[] = [];

    for (const fn of this.subscribers) {
      try {
        fn(arg);
      } catch (e) {
        errors.push(e);
        if (this.isBailMode) {
          break;
        }
      }
    }

    return errors;
  }

  for(filter: Filter<TArg>): Emitter<TArg> {
    return new Emitter<TArg>({ source: this, filter, bail: this.isBailMode });
  }
}
