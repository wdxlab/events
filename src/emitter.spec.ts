import { Emitter } from './emitter';

type EventArg = { foo: string };

test('add the listeners and emit', () => {
  const emitterArg: EventArg = { foo: 'hello' };
  const emitter = new Emitter<EventArg>();
  const handler = jest.fn();
  const handler2 = jest.fn();

  emitter.on(handler);
  emitter.on(handler);
  emitter.on(handler2);
  expect(emitter.hasSubscriptions()).toBe(true);
  emitter.emit(emitterArg);

  expect(handler).toBeCalledTimes(1);
  expect(handler).toBeCalledWith(emitterArg);
  expect(handler2).toBeCalledTimes(1);
  expect(handler2).toBeCalledWith(emitterArg);
});

test('add listener once', () => {
  const emitterArg: EventArg = { foo: 'hello' };
  const emitter = new Emitter<EventArg>();
  const handler = jest.fn();
  const handler2 = jest.fn();

  emitter.once(handler);
  const onceHandler = emitter.once(handler2);

  emitter.off(onceHandler);

  emitter.emit(emitterArg);
  emitter.emit(emitterArg);

  expect(handler).toBeCalledTimes(1);
  expect(handler).toBeCalledWith(emitterArg);
  expect(handler2).not.toHaveBeenCalled();
});

test('remove listeners', () => {
  const emitterArg: EventArg = { foo: 'hello' };
  const emitter = new Emitter<EventArg>();
  const handler = jest.fn();
  const handler2 = jest.fn();

  emitter.on(handler);
  emitter.on(handler2);
  expect(emitter.hasSubscriptions()).toBe(true);
  emitter.emit(emitterArg);
  emitter.off(handler);
  expect(emitter.hasSubscriptions()).toBe(true);
  emitter.emit(emitterArg);

  expect(handler).toBeCalledTimes(1);
  expect(handler).toBeCalledWith(emitterArg);
  expect(handler2).toBeCalledTimes(2);
  expect(handler2).toBeCalledWith(emitterArg);
});

test('clear listeners', () => {
  const emitterArg: EventArg = { foo: 'hello' };
  const emitter = new Emitter<EventArg>();
  const handler = jest.fn();
  const handler2 = jest.fn();

  emitter.on(handler);
  emitter.on(handler2);
  emitter.emit(emitterArg);
  emitter.clear();
  expect(emitter.hasSubscriptions()).toBe(false);
  emitter.emit(emitterArg);

  expect(handler).toBeCalledTimes(1);
  expect(handler).toBeCalledWith(emitterArg);
  expect(handler2).toBeCalledTimes(1);
  expect(handler2).toBeCalledWith(emitterArg);
});

test('modifying on emit', () => {
  const emitterArg: EventArg = { foo: 'hello' };
  const emitter = new Emitter<EventArg>();
  const handler = jest.fn(() => {
    emitter.off(handler);
  });
  const handler2 = jest.fn();

  emitter.on(handler);
  emitter.on(handler2);
  emitter.emit(emitterArg);
  emitter.clear();
  expect(emitter.hasSubscriptions()).toBe(false);
  emitter.emit(emitterArg);

  expect(handler).toBeCalledTimes(1);
  expect(handler).toBeCalledWith(emitterArg);
  expect(handler2).toBeCalledTimes(1);
  expect(handler2).toBeCalledWith(emitterArg);
});

describe('errors', () => {
  test('non-bail', () => {
    const emitterArg: EventArg = { foo: 'hello' };
    const emitter = new Emitter<EventArg>();
    const handler1 = jest.fn();
    const handler2 = jest.fn(() => {
      throw new Error('Foo');
    });
    const handler3 = jest.fn();

    emitter.on(handler1);
    emitter.on(handler2);
    emitter.on(handler3);

    const errors = emitter.emit(emitterArg);

    expect(errors).toHaveLength(1);
    expect(errors[0]).toBeInstanceOf(Error);
    expect(handler1).toBeCalledTimes(1);
    expect(handler2).toBeCalledTimes(1);
    expect(handler3).toBeCalledTimes(1);
  });

  test('bail', () => {
    const emitterArg: EventArg = { foo: 'hello' };
    const emitter = new Emitter<EventArg>({ bail: true });
    const handler1 = jest.fn();
    const handler2 = jest.fn(() => {
      throw new Error('Foo');
    });
    const handler3 = jest.fn();

    emitter.on(handler1);
    emitter.on(handler2);
    emitter.on(handler3);

    const errors = emitter.emit(emitterArg);

    expect(errors).toHaveLength(1);
    expect(errors[0]).toBeInstanceOf(Error);
    expect(handler1).toBeCalledTimes(1);
    expect(handler2).toBeCalledTimes(1);
    expect(handler3).toBeCalledTimes(0);
  });
});

test('readonly mode', () => {
  const source = new Emitter<number>();
  const target = new Emitter<number>({ source });
  const handler1 = jest.fn();
  const handler2 = jest.fn(() => {
    throw new Error('Foo');
  });
  const handler3 = jest.fn();

  target.on(handler1);
  target.on(handler2);
  target.on(handler3);

  expect(() => target.emit(123)).toThrow('Emitter is readonly');

  const errors = source.emit(456);

  expect(errors).toHaveLength(1);
  expect(errors[0]).toBeInstanceOf(Error);
  expect(handler1).toBeCalledTimes(1);
  expect(handler2).toBeCalledTimes(1);
  expect(handler3).toBeCalledTimes(1);
});

test('filterable', () => {
  const emitter = new Emitter<{ name: string }>();
  const aaaFilter = emitter.for((arg) => arg.name === 'aaa');
  const bbbFilter = emitter.for((arg) => arg.name === 'bbb');
  const fooFilter = emitter.for((arg) => arg.name === 'foo');
  const handler1 = jest.fn();
  const handler2 = jest.fn();

  emitter.on(handler1);
  aaaFilter.on(handler2);
  bbbFilter.on(handler2);
  fooFilter.on(handler2);

  emitter.emit({ name: 'ccc' });
  emitter.emit({ name: 'ddd' });
  emitter.emit({ name: 'foo' });

  expect(handler1.mock.calls).toMatchSnapshot();
  expect(handler2.mock.calls).toMatchSnapshot();
});
