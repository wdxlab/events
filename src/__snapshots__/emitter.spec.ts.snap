// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterable 1`] = `
[
  [
    {
      "name": "ccc",
    },
  ],
  [
    {
      "name": "ddd",
    },
  ],
  [
    {
      "name": "foo",
    },
  ],
]
`;

exports[`filterable 2`] = `
[
  [
    {
      "name": "foo",
    },
  ],
]
`;
